### PR TITLE
feat: mapped accounts

### PR DESCRIPTION
**Pull request by aljo242**
## 1. Summary
Fixes QCK-452

## 2.Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 5. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

File: x/interchainstaking/keeper/address_map_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+const (
+	testChainID = "test-1"
+)
+
+// generateRandomBytes returns securely generated random bytes.
+// It will return an error if the system's secure random
+// number generator fails to function correctly, in which
+// case the caller should not continue.
+func generateRandomBytes(n int) ([]byte, error) {
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/address_map_test.go:15]: Can we make this part of utils package so it can be reused?

**Review by joe-bowman — COMMENTED** [review]
Looks good so far;

xcclookup will require a new query type to return map of addresses, indexed by chain ID (I don't think we need them bech32ified, because the first thing we do in xcclookup is decode the address anyway) for a given address:

/quicksilver/interchainstaking/v1/mapped_addresses/quick1abcdef

which is why we prefix the local address mapping by address and not chain, so we prefix iterator over it for this query :)

_if we do, we will need to add the account prefix to the ProtocolDataConnection structs_

File: x/interchainstaking/keeper/address_map_test.go (hunk 1 above)
Comment (aljo242) [reply x/interchainstaking/keeper/address_map_test.go]: https://github.com/ingenuity-build/quicksilver/pull/436/commits/e0c0614a7bc175af16ee0d302958b89021e8552a

Comment (ajansari95) [reply x/interchainstaking/keeper/address_map_test.go]: https://github.com/ingenuity-build/quicksilver/pull/443

File: x/interchainstaking/keeper/receipt.go [hunk 2]
```diff
@@ -149,7 +167,13 @@ func (k *Keeper) MintQAsset(ctx sdk.Context, sender sdk.AccAddress, senderAddres
...
 		)
-	} else {
+	case mappedAddress != nil:
+		// set mapped account
+		k.SetAddressMapPair(ctx, sender, mappedAddress, zone.ChainId)
+
+		// set send to mapped account
+		err = k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, mappedAddress, qAssets)
+	default:
 		err = k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sender, qAssets)
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/receipt.go:177]: @aljo242 @ajansari95 

What happens here if the account was previously mapped for this chain and but no memo was send this time around?

Conversely, what happens if no memo was sent and there is _NO_ mapping?

I think `default` here is:
```
  if zone.Is118() {
    mappedAddress, found := k.GetAddressMapPair(ctx, sender, zone.ChainId)
    if found {
      err = k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, mappedAddress, qAssets)
    } else {
      // burn qAssets and return funds (ICA MsgSend back to sender)
      // perhaps work this out before the minting phase?
    }
  } else {
    err = k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sender, qAssets)
  }
```

Comment (aljo242) [reply x/interchainstaking/keeper/receipt.go:210]: https://github.com/ingenuity-build/quicksilver/pull/436/commits/be99f7d8391259c0431da2c2da1a470423c22c5a
https://github.com/ingenuity-build/quicksilver/pull/436/commits/73e36c52c38a67ee0dce548c9322b40a0e15ea8a

File: x/interchainstaking/keeper/receipt.go [hunk 3]
```diff
@@ -102,8 +119,52 @@ func (k *Keeper) HandleReceiptForTransaction(ctx sdk.Context, txr *sdk.TxRespons
...
+// The function handles the following cases:
+//  1. If the zone is labeled "return to sender" or the Tx memo contains "return to sender" flag:
+//     - Mint QAssets and IBC transfer to the corresponding zone acc
+//  2. If there is no mapped account but the zone is labeled as non-118 coin type:
+//     - Do not mint QAssets and refund assets
+//  3. If a mapped account is set for a non-118 coin type zone:
+//     - Mint QAssets and send to corresponding mapped address
+//  4. If a new mapped account is provided to the function and the zone is labeled as non-118 coin type:
+//     - Mint QAssets, set new mapping for the mapped account in the keeper, and send to corresponding mapped account.
+//  5. If the zone is non-118 and no other flags are set:
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/receipt.go:165]: This should read 'if the zone _is_ 118' right?

File: x/interchainstaking/keeper/receipt.go [hunk 4]
```diff
@@ -114,43 +175,40 @@ func (k *Keeper) MintQAsset(ctx sdk.Context, sender sdk.AccAddress, senderAddres
...
 	}
 
+	// check if a remote address exists for a non 118 coin type zone
+	setMappedAddress := true
+	if mappedAddress == nil && !zone.Is_118 {
+		var found bool
+		mappedAddress, found = k.GetRemoteAddressMap(ctx, sender, zone.ChainId)
+		if !found {
+			// if not found, skip minting and refund assets
+			return k.SendTokenIBC(ctx, k.AccountKeeper.GetModuleAddress(types.ModuleName), senderAddress, zone, assets[0])
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/receipt.go:185]: The assets in this case are still on the host zone, and should be sent using MsgSend from the zone.DepositAccount to the sender.

File: x/interchainstaking/keeper/receipt.go [hunk 5]
```diff
@@ -114,43 +175,40 @@ func (k *Keeper) MintQAsset(ctx sdk.Context, sender sdk.AccAddress, senderAddres
 		qAssets = qAssets.Add(sdk.NewCoin(zone.LocalDenom, amount))
 	}
 
+	// check if a remote address exists for a non 118 coin type zone
+	setMappedAddress := true
+	if mappedAddress == nil && !zone.Is_118 {
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/receipt.go:180]:
```
&& !zone.ReturnToSender && !memoRTS
```

Do we need to include this here? non-118 + no mapped account is okay if we RTS

**Review by joe-bowman — CHANGES_REQUESTED** [review]
a couple of nits remaining

**Review by ajansari95 — DISMISSED** [review]
lgtm

File: x/interchainstaking/types/zones_test.go [hunk 6]
```diff
@@ -145,40 +148,75 @@ func TestBase64MemoToIntent(t *testing.T) {
...
 			name:    "invalid length",
 			memo:    "ToS/+Ex92tEcuMBzhukZKMVnXKS8NKaQBJTx9zza4rrxyLiP9fwLijOcPK/59acWzdcBME6ub8f0LID97qWECuxJKXmxBM1YBQyWHSAXDiwmMY78K",
-			amount:  10,
+			wantErr: true,
+		},
+		{
+			name:    "invalid base64",
+			memo:    "\xFF",
+			wantErr: true,
+		},
```
Comment (joe-bowman) [inline x/interchainstaking/types/zones_test.go:180]: This appears to duplicate the test below.

Comment (joe-bowman) [reply x/interchainstaking/types/zones_test.go:180]:
```suggestion
```

File: x/interchainstaking/types/zones_test.go [hunk 7]
```diff
@@ -145,40 +148,70 @@ func TestBase64MemoToIntent(t *testing.T) {
...
+func TestGenMemos(t *testing.T) {
+	oldMemo := "WoS/+Ex92tEcuMBzhukZKMVnXKS8bqaQBJTx9zza4rrxyLiP9fwLijOc"
+	memoBytes, err := base64.StdEncoding.DecodeString(oldMemo)
+	require.NoError(t, err)
+	memoBytes = append(memoBytes, 255, 0, 2, 1, 2)
+
+	newMemo := base64.StdEncoding.EncodeToString(memoBytes)
+	fmt.Println(newMemo)
+}
+
```
Comment (joe-bowman) [inline x/interchainstaking/types/zones_test.go:212]: Appears left over from generating the new base64 val for the above test ;)
```suggestion
```

**Review by joe-bowman — DISMISSED** [review]
lgtm

**Review by ajansari95 — APPROVED** [review]
lgtm

File: x/interchainstaking/types/zones_test.go [hunk 8]
```diff
@@ -1,6 +1,8 @@
 package types_test
 
 import (
+	"encoding/base64"
+	"fmt"
```
Comment (ajansari95) [inline x/interchainstaking/types/zones_test.go:5]:
```suggestion
```

File: x/interchainstaking/types/zones_test.go [hunk 9]
```diff
@@ -1,6 +1,7 @@
 package types_test
 
 import (
+	"encoding/base64"
```
Comment (joe-bowman) [inline x/interchainstaking/types/zones_test.go:4]:
```suggestion
```

[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/types/zones_test.go]: race condition :)

**Review by ajansari95 — APPROVED** [review]
lgtm

**Pull request merged**